Add flood fill subroutine to mpas_li_calving.F

trhille opened the pull request:
This PR moves the flood-fill code from remove_icebergs into its own generalized subroutine called flood_fill. The seed and grow masks are defined in the calling routine and passed to the flood fill routine. It is currently called by remove_icebergs and apply_calving_damage_threshold. In the future, it should be called by von Mises and float-kill calving routines in order to avoid calving over subglacial lakes.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -142,7 +143,7 @@ subroutine li_calve_ice(domain, err)
       call mpas_pool_get_config(liConfigs, 'config_print_calving_info', config_print_calving_info)
       call mpas_pool_get_config(liConfigs, 'config_data_calving', config_data_calving)
-
+      call mpas_pool_get_config(liConfigs, 'config_number_of_blocks', config_number_of_blocks)
```
```suggestion
      call mpas_pool_get_config(liConfigs, 'config_number_of_blocks', config_number_of_blocks)
      
```

trhille replied:
Fixed in commit 370583d.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -213,7 +214,12 @@ subroutine li_calve_ice(domain, err)
-
+          ! Does not currently support multiple blocks per processor.
+          if (config_number_of_blocks /= 0) then
+             call mpas_log_write("External velocity solvers require that config_number_of_blocks=0", MPAS_LOG_ERR)
```
Is this restriction the flood fill or specific to damage calving?  The error message should be adjusted accordingly.  If the restriction applies to the flood fill, it would be better to put the check there.

Also, the endif line is indented too far.

trhille replied:
I don't know the answer to this question because I don't fully understand block loops.. The flood_fill subroutine contains a block loop, but it is also called from within a block loop when using damage calving (damage_calving calls apply_damage_calving_threshold from within a block loop, which in turn calls flood_fill). My understanding was that this specific configuration would not work with config_number_of_blocks /= 0.

matthewhoffman replied:
Fair enough.  I like that you put the error message even though we never intend to run on multiple blocks.  My main point is that it is not the velocity solver but the calving routine that should be mentioned in the message.  I suggest changing it to "damagecalving requires that...".

trhille replied:
Changed in commit 12a4159

matthewhoffman commented on mode_forward/mpas_li_calving.F:
```diff
@@ -3272,14 +3272,14 @@ subroutine remove_icebergs(domain)
       call mpas_timer_start("iceberg detection")
       call mpas_log_write("Iceberg-detection flood-fill begin.")
-      ! Allocate needed scratch fields
+            ! Allocate needed scratch fields
```
Was this indentation accidental?

trhille replied:
Fixed in commit 370583d.

matthewhoffman commented on Registry.xml:
```diff
@@ -1521,7 +1521,16 @@ is the value of that variable from the *previous* time level!
 	       persistence="scratch"
-               />
+	       />
+       <var name="seedMask" type="integer" dimensions="nCells" units="none"
```
Was there a need to switch from allocatable arrays to scratch arrays?  At this point either is fine to use, but given you already had allocatables working, I'm curious why you bothered to switch it.

trhille replied:
My understanding was that it would be necessary to use scratch arrays for halo updates.

matthewhoffman replied:
Oh, that's right.  Thanks for reminding me!

matthewhoffman reviewed (changes requested):
I made a couple review comments, but Github seems to be confused about where they apply.  It looks like you have an accidental or temporary merge commit at the head of your branch.  Can you try removing that and force pushing?  I'll take a closer look after that.  Also, can you add a comment describing what testing you've done?

trhille commented:
I tested remove_icebergs by populating icebergs in the Humboldt_1to10km_r02_20210304 mesh and using these namelist settings:
&calving
    config_calving = 'thickness_threshold'
    config_calving_thickness = 11.0
    config_data_calving = .false.
    config_calving_timescale = 0.0
    config_restore_calving_front = .false.
    config_remove_icebergs = .true.
    config_remove_small_islands = .true.
    
Initial condition:
![image](https://user-images.githubusercontent.com/17446278/123478572-db9d0200-d5bc-11eb-8796-eda762027bac.png)
After 1 timestep:
![image](https://user-images.githubusercontent.com/17446278/123478599-e3f53d00-d5bc-11eb-9908-6b62bccfc052.png)

from log.landice.0000.out:
 Initial timestep 2007-07-01_00:00:00
 Generating new external velocity solver FEM grid.
 Beginning velocity solve using external velocity solver.
 Completed velocity solve using external velocity solver.
 Starting timestep number 1
 Calculating N assuming perfect ocean connection.
 Iceberg-detection flood-fill begin.
 Iceberg-detection flood-fill: allocated.
 Iceberg-detection flood-fill: updated masks.
 Initialized 101 cells to local mask
 Iceberg-detection flood-fill initialization complete.
 Flood-fill: allocated.
 Iceberg-detection flood-fill: updated masks.
 Initialized 101 cells to local mask
 Iceberg-detection flood-fill initialization complete.
 Initialized 11072 cells to global mask
   Starting global processor loop 1
     Starting local cell loop 1
     Added 0 new cells to local mask
   Added 161 new cells to global mask
   Starting global processor loop 2
     Starting local cell loop 1
     Added 0 new cells to local mask
   Added 8 new cells to global mask
   Starting global processor loop 3
     Starting local cell loop 1
     Added 0 new cells to local mask
   Added 0 new cells to global mask
 Iceberg-detection flood-fill complete. Removed 62 iceberg cells.

trhille commented:
I then tested damagecalving by assigning damage=1.0 for all cells belonging to two procs. One damage region crosses the ice front and the other does not (black is thickness == 1 contour, includes icebergs):
![image](https://user-images.githubusercontent.com/17446278/123479422-2408ef80-d5be-11eb-9b2e-0c7c94ce01c5.png)
I ran forward with these namelist settings:
&calving
    config_calving = 'damagecalving'
    config_damage_calving_method = 'threshold'
    config_preserve_damage = .true.
    config_finalize_damage_after_advection = .false.
    config_damage_calving_threshold = 0.5
    config_data_calving = .false.
    config_calving_timescale = 0.0
    config_restore_calving_front = .false.
    config_remove_icebergs = .false.
    config_remove_small_islands = .false.
    
  Initial condition (white contours damage == 1.0):
 
![image](https://user-images.githubusercontent.com/17446278/123479643-78ac6a80-d5be-11eb-9bef-9f647be6f700.png)

After 1 timestep:
![image](https://user-images.githubusercontent.com/17446278/123479676-85c95980-d5be-11eb-8078-b7223587943f.png)

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -3265,22 +3243,38 @@ subroutine remove_icebergs(domain)
+      call mpas_pool_get_field(geometryPool, 'contiguousDynamicIceOld', contiguousDynamicIceOldField)
       call mpas_allocate_scratch_field(contiguousDynamicIceOldField, single_block_in = .false.)
+      call mpas_pool_get_dimension(meshPool, 'nCells', nCells)
+      !allocate(seedMask(nCells+1))
```
Remove these 2 comments.

trhille replied:
removed

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -3314,6 +3309,154 @@ subroutine remove_icebergs(domain)
+            if (contiguousDynamicIce(iCell) == 0 .and. li_mask_is_floating_ice(cellMask(iCell))) then
+               calvingThickness(iCell) = calvingThickness(iCell) + thickness(iCell)  ! remove any remaining ice here
+               thickness(iCell) = 0.0_RKIND
+               contiguousDynamicIceOld(iCell) = 1 ! debug: make this a mask of where icebergs were removed
```
Maybe best to comment this line.

trhille replied:
commented

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -3314,6 +3309,154 @@ subroutine remove_icebergs(domain)
+         enddo
+         localIcebergCellCount = sum(contiguousDynamicIceOld)
+
+!         where(contiguousDynamicIce == 0 .and. li_mask_is_dynamic_ice(cellMask))
```
Looks like there is no reason to keep this version around.

trhille replied:
deleted

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -3314,6 +3309,154 @@ subroutine remove_icebergs(domain)
+      call mpas_deallocate_scratch_field(contiguousDynamicIceOldField, single_block_in=.false.)
+      call mpas_deallocate_scratch_field(seedMaskField, single_block_in=.false.)
+      call mpas_deallocate_scratch_field(growMaskField, single_block_in=.false.)
+      !deallocate(seedMask)
```
delete these lines too.

trhille replied:
deleted

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -3314,6 +3309,154 @@ subroutine remove_icebergs(domain)
+!
+!    routine flood_fill
+!
+!> \brief  Flood fill routine to differentiate ice sheet from icebergs
```
Please elaborate on how to use the seedMask and growMask fields in a general application.

trhille replied:
Added in commit 12a4159.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -3314,6 +3309,154 @@ subroutine remove_icebergs(domain)
+      type (mpas_pool_type), pointer :: geometryPool
+      type (mpas_pool_type), pointer :: velocityPool
+
+      type (field1dInteger), pointer :: contiguousDynamicIceField
```
This and the next field are no longer used, so please delete these lines and the associated allocate and deallocate calls.

trhille replied:
Removed

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -3314,6 +3309,154 @@ subroutine remove_icebergs(domain)
+      type (field1dInteger), pointer :: contiguousDynamicIceField
+      type (field1dInteger), pointer :: contiguousDynamicIceOldField
+
+      real (kind=RKIND), dimension(:), pointer :: calvingThickness    !thickness of ice that calves (computed in this subroutine)
```
calvingThickness and thickness are also no longer used, so zap 'em.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -3314,6 +3309,154 @@ subroutine remove_icebergs(domain)
+
+      real (kind=RKIND), dimension(:), pointer :: calvingThickness    !thickness of ice that calves (computed in this subroutine)
+      real (kind=RKIND), dimension(:), pointer :: thickness
+      integer, dimension(:), pointer :: cellMask
```
Same with cellMask

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -3314,6 +3309,154 @@ subroutine remove_icebergs(domain)
+      integer :: newMaskCountLocal, newMaskCountLocalAccum, newMaskCountGlobal
+      integer :: err_tmp, err
+      integer :: globalLoopCount, localLoopCount
+      integer :: localIcebergCellCount, globalIcebergCellCount
```
These counters not used either.

trhille replied:
These counters are in fact still used.

matthewhoffman reviewed (changes requested):
@trhille , thanks for updating this PR.  I made a number of new comments - mostly on removing variables and lines of code that are now vestigial.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -3244,9 +3218,13 @@ subroutine remove_icebergs(domain)
       type (mpas_pool_type), pointer :: velocityPool
+      type (mpas_pool_type), pointer :: scratchPool
 
       type (field1dInteger), pointer :: contiguousDynamicIceField
```
Can you remove these two Field variables declarations?

Also it looks this routine no longer needs contiguousDynamicIce and contiguousDynamicIceOld at all and we can completely remove from this routine and then from Registry.  In the part where icebergs get removed, we can just use the udpated seedMask.

trhille replied:
contiguousDynamicIce and contiguousDynamicIceOld have been completely removed in commit 8763f9f and replaced by seedMask and seedMaskOld, as appropriate. Testing shows the correct behavior with both remove_icebergs and damagecalving, as in previous tests shown above.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -3314,6 +3303,136 @@ subroutine remove_icebergs(domain)
+!> The calling routine defines the seedMask to specify the initial masked region, and the growMask 
+!> to specify the potential region to be filled by the flood-fill routine. The flood-fill routine
+!> then adds the contiguous areas of the growMask to the seedMask to produce the final mask, which
+!> is passed back to the calling routine for application.
```
Can you clarify that the result is passed back in the seedMask variable?  (That's a little confusing as a naming convention, but I can't think of a more general but still descriptive name, and I'm not sure I like the idea of needing a third variable for the result.)

And add that any integer fields in Registry can be used for seedMask and growMask but there are existing variables with those names that can be used.

trhille replied:
Amended in commit 8763f9f.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_calving.F:
```diff
@@ -3359,24 +3478,24 @@ subroutine remove_icebergs(domain)
-               contiguousDynamicIceOld(:) = contiguousDynamicIce(:)
+               seedMaskOld(:) = seedMask(:)
 
                do iCell = 1, nCellsArray(1) ! this gives owned cells only
```
Can you change nCellsArray(1) to nCellsSolve and remove all usage of nCellsArray in this routine?

trhille replied:
Amended in commit 8763f9f.

matthewhoffman reviewed (changes requested):
@trhille , I have more nitpicky comments.  Sorry to not notice these things last time.

stephenprice commented:
This looks awesome! Great work you two.
